src/category_theory/category/Bipointed.lean

```diff
@@ -0,0 +1,125 @@
+def of {X : Type*} (to_prod : X × X) : Bipointed := ⟨X, to_prod⟩
+
+alias of ← prod.Bipointed
+
+instance : inhabited Bipointed := ⟨of ((), ())⟩
+
+/-- Morphisms in `Bipointed`. -/
+@[ext] protected structure hom (X Y : Bipointed.{u}) : Type u :=
+(to_fun : X → Y)
+(map_fst : to_fun X.to_prod.1 = Y.to_prod.1)
+(map_snd : to_fun X.to_prod.2 = Y.to_prod.2)
+
+namespace hom
+
+/-- The identity morphism of `X : Bipointed`. -/
+@[simps] def id (X : Bipointed) : hom X X := ⟨id, rfl, rfl⟩
+
+instance (X : Bipointed) : inhabited (hom X X) := ⟨id X⟩
+
+/-- Composition of morphisms of `Bipointed`. -/
+@[simps] def comp {X Y Z : Bipointed.{u}} (f : hom X Y) (g : hom Y Z) : hom X Z :=
+⟨g.to_fun ∘ f.to_fun, by rw [function.comp_apply, f.map_fst, g.map_fst],
+  by rw [function.comp_apply, f.map_snd, g.map_snd]⟩
+
+end hom
+
+instance large_category : large_category Bipointed :=
+{ hom := hom,
+  id := hom.id,
+  comp := @hom.comp,
+  id_comp' := λ _ _ _, hom.ext _ _ rfl,
+  comp_id' := λ _ _ _, hom.ext _ _ rfl,
+  assoc' := λ _ _ _ _ _ _ _, hom.ext _ _ rfl }
+
+instance concrete_category : concrete_category Bipointed :=
+{ forget := { obj := Bipointed.X, map := @hom.to_fun },
+  forget_faithful := ⟨@hom.ext⟩ }
+
+/-- Swaps the pointed elements of a bipointed type. `prod.swap` as a functor. -/
+@[simps] def swap : Bipointed ⥤ Bipointed :=
+{ obj := λ X, ⟨X, X.to_prod.swap⟩, map := λ X Y f, ⟨f.to_fun, f.map_snd, f.map_fst⟩ }
+
+/-- The equivalence between `Bipointed` and itself induced by `prod.swap` both ways. -/
+def swap_equiv : Bipointed ≌ Bipointed :=
+equivalence.mk swap swap
+  (nat_iso.of_components (λ X, { hom := ⟨id, rfl, rfl⟩, inv := ⟨id, rfl, rfl⟩ }) $ λ X Y f, rfl)
+  (nat_iso.of_components (λ X, { hom := ⟨id, rfl, rfl⟩, inv := ⟨id, rfl, rfl⟩ }) $ λ X Y f, rfl)
+
+@[simp] lemma swap_equiv_symm : swap_equiv.symm = swap_equiv := rfl
+
+end Bipointed
+
+/-- The forgetful functor from `Bipointed` to `Pointed` which forgets about the second point. -/
+def Bipointed_to_Pointed_fst : Bipointed ⥤ Pointed :=
+{ obj := λ X, ⟨X, X.to_prod.1⟩, map := λ X Y f, ⟨f.to_fun, f.map_fst⟩ }
+
+/-- The forgetful functor from `Bipointed` to `Pointed` which forgets about the first point. -/
+def Bipointed_to_Pointed_snd : Bipointed ⥤ Pointed :=
+{ obj := λ X, ⟨X, X.to_prod.2⟩, map := λ X Y f, ⟨f.to_fun, f.map_snd⟩ }
+
+@[simp] lemma Bipointed_to_Pointed_fst_comp_forget :
+  Bipointed_to_Pointed_fst ⋙ forget Pointed = forget Bipointed := rfl
+
+@[simp] lemma Bipointed_to_Pointed_snd_comp_forget :
+  Bipointed_to_Pointed_snd ⋙ forget Pointed = forget Bipointed := rfl
+
+@[simp] lemma swap_comp_Bipointed_to_Pointed_fst :
+  Bipointed.swap ⋙ Bipointed_to_Pointed_fst = Bipointed_to_Pointed_snd := rfl
+
+@[simp] lemma swap_comp_Bipointed_to_Pointed_snd :
+  Bipointed.swap ⋙ Bipointed_to_Pointed_snd = Bipointed_to_Pointed_fst := rfl
+
+/-- The functor from `Pointed` to `Bipointed` which adds a second point. -/
+def Pointed_to_Bipointed_fst : Pointed.{u} ⥤ Bipointed :=
```

Comment:
I guess you want to `TODO` the adjunction of this with `Bipointed_to_Pointed_fst`?